### PR TITLE
New Feature: KDBX4

### LINKED ISSUE #148: Adding support for KDBX4 file format
> KeePass 2.35 introduces version 4 of the KDBX file format. This new format features both improvements and new capabilities. These are outlined in the following sections.

http://keepass.info/help/kb/kdbx_4.html

### LINKED ISSUE #1060: Make implementation of Challenge-Response more compatible to Keepass 2
As the author of Keepass2Android I was looking into [implementing the Challenge-Response functionality as in KeepassXC](https://github.com/PhilippC/keepass2android/issues/4). While I like the idea of using the Master seed as "challenge", I am not so happy with the implementation of Challenge-Response. To explain why, here is a short overview of the key generation in Keepass 2 (by Dominik Reichl) which is used by Keepass2Android as well:

Keepass2:
```
pbPassword = CryptoUtil.HashSha256(StrUtil.Utf8.GetBytes(strPassword))
pbKeyFile = LoadKeyFile(...)
pbCustomKey = challengeResponse //for example when using KeeChallenge Plugin in keepass 2
//CompositeKey concatenates the data of ALL IUserKey objects
pbAllData = pbPassword + pbKeyfile + pbCustomKey
pbRaw32 = CryptoUtil.HashSha256(pbAllData)
pUserKey32 = kdf.Transform(pbRaw32)
pbCmp = MasterSeed + pUserKey32
pbHmacKey64 = h.ComputeHash(pbCmp);
```

KeepassXC, with the naming conventions from Keepass 2
```
pbPassword = CryptoUtil.HashSha256(StrUtil.Utf8.GetBytes(strPassword))
pbKeyFile = LoadKeyFile(...)
//CompositeKey only concatenates password + keyfile but not challengeResponse
pbAllData = pbPassword + pbKeyFile
pUserKey32 = kdf.Transform(pbRaw32)
pbRaw32 = CryptoUtil.HashSha256(pbAllData)
//challengeResponse is added later, where no plugin can modify the computation in Keepass2
pbCmp = MasterSeed + challengeResponse + pUserKey32
pbHmacKey64 = h.ComputeHash(pbCmp);
```

In Keepass 2, it is very common for Plugins to add elements to the Composite Key. These elements are objects of `KcpCustomKey` or custom classes implementing the `IUserKey` interface. The data of all these elements are concatenated and transformed. When I added support for KeeOtp or KeeChallenge in Keepass2Android, I did exactly this, which was very simple.

In KeepassXC, the challengeRespose is not concatenated with password and (potentially) key file. Instead, it is prepended to the hashed and transformed key. This cannot be implemented for Keepass2 by a plugin, so the current KeepassXC implementation will not be usable for Keepass 2 users, and nobody can change this by developing a plugin. It also requires some pretty intrusive code changes if I want to implement this in Keepass2Android which I believe are not necessary if KeepassXC would offer (at least as an option or better as the default) to concat the challengeResponse to the pbAllData.

If I didn't miss any important reason to integrate the challengeResponse at this point in the processing, I suggest you make the key computation more compatible to other Keepass implementations. Please let me know what you think so I can plan further development on the feature for keepass2android.

### PR DESCRIPTION
**Pull request by phoerious**
Working PR for KDBX4. Follows up on #1179.

Closes #148 and closes #1060 

## Types of changes

- ✅ New feature (non-breaking change which adds functionality)
- ✅ Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- ✅ I have read the **CONTRIBUTING** document. **[REQUIRED]**
- ✅  My code follows the code style of this project. **[REQUIRED]**
- ✅ All new and existing tests passed. **[REQUIRED]**
- ✅ I have compiled and verified my code with `-DWITH_ASAN=ON`. **[REQUIRED]**
- ✅  I have added tests to cover my changes.

### CODE REVIEW

**Comment by angelsl** [comment]
Be sure to pick up the [rest of the KDBX 4 changes](https://github.com/angelsl/keepassxc/tree/kdbx42) as well.

**Comment by droidmonkey** [comment]
Reminder to implement #1060

**Comment by phoerious** [comment]
Aye.

File: src/format/Kdbx3Reader.cpp [hunk 1]
```diff
@@ -0,0 +1,376 @@
...
+        buffer->open(QIODevice::ReadOnly);
+        xmlDevice = buffer.data();
+    }
+
+    Kdbx3XmlReader xmlReader;
+    xmlReader.readDatabase(xmlDevice, m_db, &randomStream);
+
+    if (xmlReader.hasError()) {
+        raiseError(xmlReader.errorString());
+        if (keepDatabase) {
```
Comment (adolfogc) [inline src/format/Kdbx3Reader.cpp]: I think this reads better as:
```cpp
return keepDatabase ? db.take() : nullptr;
```

File: src/format/Kdbx3Reader.cpp [hunk 2]
```diff
@@ -0,0 +1,376 @@
...
+    }
+
+    HashedBlockStream hashedStream(&cipherStream);
+    if (!hashedStream.open(QIODevice::ReadOnly)) {
+        raiseError(hashedStream.errorString());
+        return nullptr;
+    }
+
+    QIODevice* xmlDevice;
+    QScopedPointer<QtIOCompressor> ioCompressor;
```
Comment (adolfogc) [inline src/format/Kdbx3Reader.cpp]: What do you think of replacing `QtIOCompressor` with `KCompressionDevice` from KF5Archive?

Comment (droidmonkey) [reply src/format/Kdbx3Reader.cpp]: What would be the purpose?

Comment (adolfogc) [reply src/format/Kdbx3Reader.cpp]: Remove the embedded version.

Comment (phoerious) [reply src/format/Kdbx3Reader.cpp]: I'm against adding new dependencies here if not really needed.

Comment (adolfogc) [reply src/format/Kdbx3Reader.cpp]: Ok; makes sense. I've also noticed there is no package available for Ubuntu Trusty. Please disregard.

File: src/format/KeePass2Writer.h [hunk 3]
```diff
@@ -18,33 +18,48 @@
...
 
+#include "core/Database.h"
 #include "format/KeePass2.h"
-#include "keys/CompositeKey.h"
 
-class Database;
-class QIODevice;
-
-class KeePass2Writer
+class BaseKeePass2Writer
```
Comment (phoerious) [inline src/format/KeePass2Writer.h]: KeePass2Writer and BaseKeePass2Writer should be merged into KeePass2Writer.

File: src/format/KeePass2Reader.cpp [hunk 4]
```diff
@@ -223,217 +55,120 @@ Database* KeePass2Reader::readDatabase(const QString& filename, const CompositeK
...
-            raiseError("Unsupported compression algorithm");
-        }
-        else {
-            m_db->setCompressionAlgo(static_cast<Database::CompressionAlgorithm>(id));
-        }
+    // Determine KDBX3 vs KDBX4
+    if (m_version < KeePass2::FILE_VERSION_4) {
+        m_reader.reset(new Kdbx3Reader());
+    } else {
+        m_reader.reset(new Kdbx4Reader());
```
Comment (phoerious) [inline src/format/KeePass2Reader.cpp]: This doesn't make sense. Don't use inheriting classes.

File: src/format/KeePass2Writer.cpp [hunk 5]
```diff
@@ -15,196 +15,75 @@
...
+
+    if (db->publicCustomData().size() > 0) {
+        useKdbx4 = true;
+    }
+
+    // Determine KDBX3 vs KDBX4
+    if (useKdbx4) {
+        m_writer.reset(new Kdbx4Writer());
+    } else {
+        m_writer.reset(new Kdbx3Writer());
```
Comment (phoerious) [inline src/format/KeePass2Writer.cpp]: Don't use classes which are further down in the inheritance hierarchy.

File: src/streams/HmacBlockStream.cpp [hunk 6]
```diff
@@ -0,0 +1,267 @@
...
+    m_error = false;
+}
+
+bool HmacBlockStream::reset()
+{
+    // Write final block(s) only if device is writable and we haven't
+    // already written a final block.
+    if (isWritable() && (!m_buffer.isEmpty() || m_blockIndex != 0)) {
+        if (!m_buffer.isEmpty()) {
+            if (!writeHashedBlock()) {
```
Comment (phoerious) [inline src/streams/HmacBlockStream.cpp]: Combine conditions.

File: src/streams/HmacBlockStream.cpp [hunk 7]
```diff
@@ -0,0 +1,267 @@
...
+
+    qint64 bytesRemaining = maxSize;
+    qint64 offset = 0;
+
+    while (bytesRemaining > 0) {
+        if (m_bufferPos == m_buffer.size()) {
+            if (!readHashedBlock()) {
+                if (m_error) {
+                    return -1;
+                } else {
```
Comment (phoerious) [inline src/streams/HmacBlockStream.cpp]: Drop else.

File: src/streams/HmacBlockStream.cpp [hunk 8]
```diff
@@ -0,0 +1,267 @@
...
+        m_buffer.append(data + offset, bytesToCopy);
+
+        offset += bytesToCopy;
+        bytesRemaining -= bytesToCopy;
+
+        if (m_buffer.size() == m_blockSize) {
+            if (!writeHashedBlock()) {
+                if (m_error) {
+                    return -1;
+                } else {
```
Comment (phoerious) [inline src/streams/HmacBlockStream.cpp]: Drop else.

File: src/streams/HmacBlockStream.cpp [hunk 9]
```diff
@@ -0,0 +1,267 @@
...
+    if (!m_buffer.isEmpty()) {
+        if (m_baseDevice->write(m_buffer) != m_buffer.size()) {
+            m_error = true;
+            setErrorString(m_baseDevice->errorString());
+            return false;
+        }
+
+        m_buffer.clear();
+    }
+    m_blockIndex++;
```
Comment (phoerious) [inline src/streams/HmacBlockStream.cpp]: ++m_blockIndex

File: src/streams/HmacBlockStream.cpp [hunk 10]
```diff
@@ -0,0 +1,267 @@
...
+    hasher.addData(m_buffer);
+
+    if (hmac != hasher.result()) {
+        m_error = true;
+        setErrorString("Mismatch between hash and data.");
+        return false;
+    }
+
+    m_bufferPos = 0;
+    m_blockIndex++;
```
Comment (phoerious) [inline src/streams/HmacBlockStream.cpp]: ++m_blockIndex

File: src/streams/SymmetricCipherStream.cpp [hunk 11]
```diff
@@ -159,26 +155,28 @@ bool SymmetricCipherStream::readBlock()
...
+                    m_error = true;
+                    setErrorString("Invalid padding.");
+                    return false;
+                } else {
+                    Q_ASSERT(m_buffer.right(padLength) == QByteArray(padLength, padLength));
+                    // resize buffer to strip padding
+                    m_buffer.resize(blockSize() - padLength);
+                    return true;
+                }
+            } else {
```
Comment (phoerious) [inline src/streams/SymmetricCipherStream.cpp]: Drop else.

File: src/streams/SymmetricCipherStream.cpp [hunk 12]
```diff
@@ -250,3 +248,11 @@ bool SymmetricCipherStream::writeBlock(bool lastBlock)
         return true;
     }
 }
+
+int SymmetricCipherStream::blockSize() const {
+    if (m_streamCipher) {
+        return 1024;
+    } else {
```
Comment (phoerious) [inline src/streams/SymmetricCipherStream.cpp]: Drop else.

File: tests/TestKeePass2Reader.cpp [hunk 13]
```diff
@@ -155,3 +155,26 @@ void TestKeePass2Reader::testFormat300()
 
     delete db;
 }
+
+void TestKeePass2Reader::testFormat400()
+{
+    QString filename = QString(KEEPASSX_TEST_DATA_DIR).append("/Format400.kdbx");
```
Comment (phoerious) [inline tests/TestKeePass2Reader.cpp]: Was this file generated by KeePass2? We should use that as reference.

File: tests/TestKeePass2XmlReader.cpp [hunk 14]
```diff
@@ -79,18 +79,82 @@ QByteArray TestKeePass2XmlReader::strToBytes(const QString& str)
...
-    KeePass2XmlReader reader;
+    Kdbx4XmlReader reader;
     reader.setStrictMode(true);
     QString xmlFile = QString(KEEPASSX_TEST_DATA_DIR).append("/NewDatabase.xml");
     m_db = reader.readDatabase(xmlFile);
     QVERIFY(m_db);
     QVERIFY(!reader.hasError());
 }
 
+void TestKdbx3XmlReader::readDatabase(QString path, bool strictMode, Database*& db, bool& hasError, QString& errorString)
```
Comment (phoerious) [inline tests/TestKeePass2XmlReader.cpp]: hasError should be the return value.

File: tests/TestKeePass2XmlReader.cpp [hunk 15]
```diff
@@ -459,19 +533,19 @@ void TestKeePass2XmlReader::testInvalidXmlChars()
...
+    writeDatabase(&buffer, dbWrite.data(), hasError, errorString);
+    QVERIFY(!hasError);
     buffer.seek(0);
 
-    KeePass2XmlReader reader;
-    reader.setStrictMode(true);
-    QScopedPointer<Database> dbRead(reader.readDatabase(&buffer));
-    if (reader.hasError()) {
-        qWarning("Database read error: %s", qPrintable(reader.errorString()));
+    Database* dbRead;
```
Comment (phoerious) [inline tests/TestKeePass2XmlReader.cpp]: You can use the dereference operator on a QScopedPointer to get a reference to the guarded raw pointer. No need to revert back here.

File: tests/TestKeePass2XmlReader.cpp [hunk 16]
```diff
@@ -486,22 +560,28 @@ void TestKeePass2XmlReader::testInvalidXmlChars()
...
 
 void TestKeePass2XmlReader::testRepairUuidHistoryItem()
 {
-    KeePass2XmlReader reader;
     QString xmlFile = QString("%1/%2.xml").arg(KEEPASSX_TEST_DATA_DIR, "BrokenDifferentEntryHistoryUuid");
     QVERIFY(QFile::exists(xmlFile));
-    QScopedPointer<Database> db(reader.readDatabase(xmlFile));
-    if (reader.hasError()) {
-        qWarning("Database read error: %s", qPrintable(reader.errorString()));
+    Database* db;
```
Comment (phoerious) [inline tests/TestKeePass2XmlReader.cpp]: Same here.

File: src/format/Kdbx4Writer.cpp [hunk 17]
```diff
@@ -0,0 +1,326 @@
...
+    SymmetricCipherStream* cipherStream = new SymmetricCipherStream(hmacStream, algo,
+                                                                    SymmetricCipher::algorithmMode(algo),
+                                                                    SymmetricCipher::Encrypt);
+    if (!cipherStream->init(finalKey, encryptionIV)) {
+        raiseError(cipherStream->errorString());
+        return false;
+    }
+    if (!cipherStream->open(QIODevice::WriteOnly)) {
+        raiseError(cipherStream->errorString());
+        return false;
```
[the remark this answers is not in the document]
Comment (phoerious) [reply src/format/Kdbx4Writer.cpp]: Fixed.

File: src/format/Kdbx4XmlReader.cpp [hunk 18]
```diff
@@ -0,0 +1,1080 @@
...
+                    bool ok;
+                    QByteArray plaintext = m_randomStream->process(ciphertext, &ok);
+                    if (!ok) {
+                        value.clear();
+                        raiseError(m_randomStream->errorString());
+                    } else {
+                        value = QString::fromUtf8(plaintext);
+                    }
+                } else {
+                    raiseError("Unable to decrypt entry string");
```
[the remark this answers is not in the document]
Comment (phoerious) [reply src/format/Kdbx4XmlReader.cpp]: Fixed.

File: src/format/Kdbx4XmlReader.cpp [hunk 19]
```diff
@@ -0,0 +1,1080 @@
...
+    QString value;
+    bool protect = false;
+    bool keySet = false;
+    bool valueSet = false;
+
+    while (!m_xml.hasError() && m_xml.readNextStartElement()) {
+        if (m_xml.name() == "Key") {
+            key = readString();
+            keySet = true;
+        } else if (m_xml.name() == "Value") {
```
[the remark this answers is not in the document]
Comment (phoerious) [reply src/format/Kdbx4XmlReader.cpp]: Fixed.

File: src/format/Kdbx4XmlReader.cpp [hunk 20]
```diff
@@ -0,0 +1,1080 @@
...
+    QString key;
+    QByteArray value;
+    bool keySet = false;
+    bool valueSet = false;
+
+    while (!m_xml.hasError() && m_xml.readNextStartElement()) {
+        if (m_xml.name() == "Key") {
+            key = readString();
+            keySet = true;
+        } else if (m_xml.name() == "Value") {
```
[the remark this answers is not in the document]
Comment (phoerious) [reply src/format/Kdbx4XmlReader.cpp]: Fixed.

File: src/streams/SymmetricCipherStream.cpp (hunk 11 above)
Comment (phoerious) [reply src/streams/SymmetricCipherStream.cpp]: Fixed.

File: src/format/KeePass2.cpp [hunk 21]
```diff
@@ -0,0 +1,113 @@
...
+{
+    if (uuid == KDF_AES) {
+        return QSharedPointer<AesKdf>::create();
+    }
+    if (uuid == KDF_ARGON2) {
+        return QSharedPointer<Argon2Kdf>::create();
+    }
+
+    Q_ASSERT_X(false, "uuidToKdf", "Invalid UUID");
+    return nullptr;
```
Comment (TheZ3ro) [inline src/format/KeePass2.cpp]: return `{}` instead

File: src/format/KeePass2.cpp [hunk 22]
```diff
@@ -0,0 +1,129 @@
...
+const QString KeePass2::KDFPARAM_ARGON2_ASSOCDATA("A");
+
+const QList<QPair<Uuid, QString>> KeePass2::CIPHERS{
+    qMakePair(KeePass2::CIPHER_AES, QObject::tr("AES: 256-bit")),
+    qMakePair(KeePass2::CIPHER_TWOFISH, QObject::tr("Twofish: 256-bit")),
+    qMakePair(KeePass2::CIPHER_CHACHA20, QObject::tr("ChaCha20: 256-bit"))
+};
+
+const QList<QPair<Uuid, QString>> KeePass2::KDFS{
+    qMakePair(KeePass2::KDF_ARGON2, QObject::tr("Argon2 (recommended)")),
```
Comment (droidmonkey) [inline src/format/KeePass2.cpp]: I think this should be "Argon2 (KDBX 4 - recommended)"

File: src/gui/DatabaseSettingsWidget.cpp [hunk 23]
```diff
@@ -51,56 +75,112 @@ void DatabaseSettingsWidget::load(Database* db)
...
+    // Setup kdf parameters
+    auto kdf = m_db->kdf();
+    m_uiEncryption->transformRoundsSpinBox->setValue(kdf->rounds());
+    if (kdfUuid == KeePass2::KDF_ARGON2) {
+        auto argon2Kdf = kdf.staticCast<Argon2Kdf>();
+        m_uiEncryption->memorySpinBox->setValue(static_cast<int>(argon2Kdf->memory()) / (1 << 10));
+        m_uiEncryption->parallelismSpinBox->setValue(argon2Kdf->parallelism());
+    }
+
+    m_uiGeneral->dbNameEdit->setFocus();
```
Comment (droidmonkey) [inline src/gui/DatabaseSettingsWidget.cpp]: Need to add ```m_ui->categoryList->setCurrentCategory(0);``` after this line to highlight the first category by default.

**Pull request merged**